### PR TITLE
Simplify RootedReference and make it specifically about slices

### PR DESCRIPTION
**Pull request by nox**
It's now called `DomSlice<T>`.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23013)

### CODE REVIEW

**Comment by nox** [comment]
r? @jdm 

I'm making use of `Option::deref` for `inert` so we may as well clean up some old stuff while at it.

File: components/script/dom/bindings/root.rs [hunk 1]
```diff
@@ -256,18 +256,23 @@ pub unsafe fn trace_roots(tracer: *mut JSTracer) {
...
+}
+
+impl<T> DomSlice<T> for [Dom<T>]
+where
+    T: JSTraceable + DomObject,
+{
+    #[inline]
+    fn r(&self) -> &[&T] {
+        let _ = mem::transmute::<Dom<T>, &T>;
+        unsafe { &*(self as *const [Dom<T>] as *const &[&T]) }
```
Comment (jdm) [inline components/script/dom/bindings/root.rs]: Why is it safe to cast `*const [Dom<T>]` to `* const &[&T]`? (note extra &)

Comment (nox) [reply components/script/dom/bindings/root.rs]: Because I can't type. Will remove the `&`. How that even type checks is a mystery to me.

**Pull request merged**